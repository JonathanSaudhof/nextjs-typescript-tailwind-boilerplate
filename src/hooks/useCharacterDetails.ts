import useSWR from "swr";
import fetcher from "@/lib/fetcher";

import { extractIdFromUrl, prefetcher } from "@/lib/helper";
import { CustomError } from "@/lib/customError";

export interface CharacterRawDetails {
  birth_year: string;
  created: string;
  edited: string;
  eye_color: string;
  films: string[];
  gender: string;
  hair_color: string;
  height: string;
  homeworld: string;
  mass: string;
  name: string;
  skin_color: string;
  species: string[];
  starships: string[];
  url: string;
  vehicles: string[];
}

interface CharacterDetails {
  name: string;
  hairColor: string;
  skinColor: string;
  birthYear: string;
  gender: string;
  films: number[];
  species: string[];
}

export function useCharacterDetails(id: number | null): {
  characterDetails: CharacterDetails | null;
  isLoading: boolean;
  isError: CustomError;
} {
  const { data, error } = useSWR<CharacterRawDetails>(
    id ? `https://swapi.dev/api/people/${id}` : null,
    fetcher,
  );

  if (data) {
    const {
      name,
      hair_color: hairColor,
      skin_color: skinColor,
      birth_year: birthYear,
      gender,
      films: rawFilms, // ids need to be extracted
      species,
    } = data;

    const films = rawFilms.map((film) => extractIdFromUrl(film));

    return {
      characterDetails: {
        name,
        hairColor,
        skinColor,
        birthYear,
        gender,
        films,
        species,
      },
      isLoading: false,
      isError: error,
    };
  }

  return {
    characterDetails: null,
    isLoading: true,
    isError: error,
  };
}

export async function prefetchCharacterDetails(id: number) {
  const url = `https://swapi.dev/api/people/?${id}`;
  prefetcher(url);
}
