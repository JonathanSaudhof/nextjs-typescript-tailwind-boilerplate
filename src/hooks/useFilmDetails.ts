import useSWR from "swr";
import fetcher from "@src/lib/fetcher";
import { prefetcher } from "@lib/helper";

interface FilmRawDetails {
  id: number;
  episodeId: string;
}

export default function useFilmDetails(id: number): {
  filmDetails: FilmRawDetails;
  isLoading: boolean;
  isError: Error;
} {
  const { data, error } = useSWR(`https://swapi.dev/api/films/${id}`, fetcher);

  if (data) {
    const { count, results: characterRawList } = data;
    numberOfPages = Math.round(+count / ITEMS_PER_PAGE);
    characterList = extractCharacterListItemInformation(characterRawList);

    return {
      characterList,
      currentPage: page,
      pages: numberOfPages,
      isLoading: false,
      isError: error,
    };
  }

  return {
    characterList: [],
    currentPage: page,
    pages: numberOfPages,
    isLoading: true,
    isError: error,
  };
}

export function prefetchFilms(id: number) {
  const url = `https://swapi.dev/api/films/${id}/`;
  prefetcher(url);
}
