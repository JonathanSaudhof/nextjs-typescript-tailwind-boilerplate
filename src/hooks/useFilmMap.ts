import useSWR from "swr";
import fetcher from "@src/lib/fetcher";
import { extractIdFromUrl } from "@lib/helper";

interface FilmRawDetails {
  episode_id: string;
  title: string;
  opening_crawl: string;
  director: string;
  producer: string;
  characters: string[];
  species: string[];
  vehicles: string[];
  starships: string[];
  planets: string[];
  release_date: string;
  created: string;
  edited: string;
  url: string;
}

interface FilmRawList {
  results: FilmRawDetails[];
  count: number;
  next: string;
  previous: string;
}

interface FilmDetails {
  title: string;
  episodeId: string;
  openingCrawl: string;
  director: string;
  producer: string;
  characters: number[];
  releaseDate: string;
}

interface FilmMap {
  [id: number]: FilmDetails;
}

export default function useFilmMap(): {
  filmMap: FilmMap | null;
  isLoading: boolean;
  isError: Error;
} {
  const { data, error } = useSWR<FilmRawList>(
    `https://swapi.dev/api/films`,
    fetcher,
  );

  if (data) {
    const { results: filmRawList } = data;

    let filmMap: FilmMap = {};

    for (const film of filmRawList) {
      const id = extractIdFromUrl(film.url);
      const characters = film.characters.map((character) =>
        extractIdFromUrl(character),
      );

      filmMap[id] = {
        title: film.title,
        episodeId: film.episode_id,
        openingCrawl: film.opening_crawl,
        director: film.director,
        producer: film.producer,
        characters,
        releaseDate: film.release_date,
      };
    }

    return {
      filmMap,
      isLoading: false,
      isError: error,
    };
  }

  return {
    filmMap: null,
    isLoading: true,
    isError: error,
  };
}
