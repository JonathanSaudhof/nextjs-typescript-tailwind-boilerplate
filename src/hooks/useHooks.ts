export function useHooks() {
  // use Hooks in here
}
