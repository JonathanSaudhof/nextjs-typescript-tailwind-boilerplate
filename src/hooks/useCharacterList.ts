import useSWR from "swr";
import fetcher from "@/lib/fetcher";
import { extractIdFromUrl, prefetcher } from "@/lib/helper";
import { CharacterRawDetails } from "./useCharacterDetails";

const ITEMS_PER_PAGE = 10;

export type CharcerRawList = {
  count: number;
  next: string;
  previous: string | null;
  results: CharacterRawDetails[];
};

interface CharcterDetails {
  id: number;
  name: string;
  gender: string;
  species: string[];
  birthYear: string;
  eyeColor: string;
  films: number[];
  url: string;
}

export function useCharacterList(page: number): {
  characterList: CharcterDetails[] | [];
  pages: number | null;
  isLoading: boolean;
  isError: Error;
} {
  const { data, error } = useSWR<CharcerRawList | undefined>(
    page ? `https://swapi.dev/api/people/?page=${page}` : null,
    fetcher,
  );

  let numberOfPages: number | null = null;
  let characterList: any[] = [];

  if (data) {
    const { count, results: characterRawList } = data;
    numberOfPages = Math.round(+count / ITEMS_PER_PAGE);
    characterList = extractCharacterListItemInformation(characterRawList);

    return {
      characterList,
      pages: numberOfPages,
      isLoading: false,
      isError: error,
    };
  }

  return {
    characterList: [],
    pages: numberOfPages,
    isLoading: true,
    isError: error,
  };
}

/**
 * extractCharacterListInformation
 * @description Get a subset of data from the raw characters list retruned from the api
 * to be shown in the list view
 */
function extractCharacterListItemInformation(
  characterRawList: CharacterRawDetails[],
): CharcterDetails[] {
  return characterRawList.map((character) => {
    const id = extractIdFromUrl(character.url);

    const films = character.films.map((film) => {
      const id = extractIdFromUrl(film);

      return id;
    });

    return {
      id,
      name: character.name,
      gender: character.gender,
      species: character.species,
      birthYear: character.birth_year,
      eyeColor: character.eye_color,
      films,
      url: character.url,
    };
  });
}

export async function prefetchCharacterList(page: number) {
  const url = `https://swapi.dev/api/people/?page=${page}`;
  prefetcher(url);
}
