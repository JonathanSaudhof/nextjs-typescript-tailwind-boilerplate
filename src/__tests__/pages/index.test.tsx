import Home from "@/src/pages";
import { cleanup, render, waitFor } from "@testing-library/react";
import { works } from "../__mocks__/mockCharacterList";
import filmMap from "../__mocks__/mockFilmMap";

describe("Should render the first page on load", () => {
  const useRouter = jest.spyOn(require("next/router"), "useRouter");

  const useCharactersList = jest.spyOn(
    require("@/hooks/useCharacterList"),
    "useCharacterList",
  );

  jest.mock("@/hooks/useFilmMap", () => ({
    useFilmMap: () => filmMap,
  }));

  afterEach(() => {
    cleanup();
    jest.clearAllMocks();
  });

  it("It should be render character list on page 1", async () => {
    const mockRouter = {
      query: {},
    };

    useRouter.mockReturnValue(mockRouter);
    useCharactersList.mockImplementation(() => works);

    const { asFragment } = render(<Home />);
    await waitFor(() => expect(useRouter).toHaveBeenCalled());
    expect(useCharactersList).toHaveBeenCalledWith(1);
    expect(asFragment()).toMatchSnapshot();
  });

  it("It should be render character list on page 2", async () => {
    const mockRouter = {
      query: {
        page: "2",
      },
    };

    useRouter.mockReturnValue(mockRouter);
    useCharactersList.mockImplementation(() => works);

    const { asFragment } = render(<Home />);
    await waitFor(() => expect(useRouter).toHaveBeenCalled());
    expect(useCharactersList).toHaveBeenCalledWith(2);
    expect(asFragment()).toMatchSnapshot();
  });
});
