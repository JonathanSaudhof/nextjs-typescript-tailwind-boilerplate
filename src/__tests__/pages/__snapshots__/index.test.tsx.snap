// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render the first page on load It should be render character list on page 1 1`] = `
<DocumentFragment>
  <div
    class="container mx-auto"
  >
    <ul
      class="flex flex-col gap-6 py-20"
    >
      <li
        class="flex flex-col shadow-md bg-gray-500 border rounded"
      >
        <a
          href="/character/1"
        >
          Luke Skywalker
        </a>
        <div>
          male
        </div>
        <div>
          https://swapi.dev/api/people/1/
        </div>
      </li>
      <li
        class="flex flex-col shadow-md bg-gray-500 border rounded"
      >
        <a
          href="/character/2"
        >
          C-3PO
        </a>
        <div>
          n/a
        </div>
        <div>
          https://swapi.dev/api/people/2/
        </div>
      </li>
      <li
        class="flex flex-col shadow-md bg-gray-500 border rounded"
      >
        <a
          href="/character/3"
        >
          R2-D2
        </a>
        <div>
          n/a
        </div>
        <div>
          https://swapi.dev/api/people/3/
        </div>
      </li>
      <li
        class="flex flex-col shadow-md bg-gray-500 border rounded"
      >
        <a
          href="/character/4"
        >
          Darth Vader
        </a>
        <div>
          male
        </div>
        <div>
          https://swapi.dev/api/people/4/
        </div>
      </li>
      <li
        class="flex flex-col shadow-md bg-gray-500 border rounded"
      >
        <a
          href="/character/5"
        >
          Leia Organa
        </a>
        <div>
          female
        </div>
        <div>
          https://swapi.dev/api/people/5/
        </div>
      </li>
      <li
        class="flex flex-col shadow-md bg-gray-500 border rounded"
      >
        <a
          href="/character/6"
        >
          Owen Lars
        </a>
        <div>
          male
        </div>
        <div>
          https://swapi.dev/api/people/6/
        </div>
      </li>
      <li
        class="flex flex-col shadow-md bg-gray-500 border rounded"
      >
        <a
          href="/character/7"
        >
          Beru Whitesun lars
        </a>
        <div>
          female
        </div>
        <div>
          https://swapi.dev/api/people/7/
        </div>
      </li>
      <li
        class="flex flex-col shadow-md bg-gray-500 border rounded"
      >
        <a
          href="/character/8"
        >
          R5-D4
        </a>
        <div>
          n/a
        </div>
        <div>
          https://swapi.dev/api/people/8/
        </div>
      </li>
      <li
        class="flex flex-col shadow-md bg-gray-500 border rounded"
      >
        <a
          href="/character/9"
        >
          Biggs Darklighter
        </a>
        <div>
          male
        </div>
        <div>
          https://swapi.dev/api/people/9/
        </div>
      </li>
      <li
        class="flex flex-col shadow-md bg-gray-500 border rounded"
      >
        <a
          href="/character/10"
        >
          Obi-Wan Kenobi
        </a>
        <div>
          male
        </div>
        <div>
          https://swapi.dev/api/people/10/
        </div>
      </li>
    </ul>
    <ul
      class="fixed bottom-0 flex gap-6 py-6 bg-gray-800 w-full"
    >
      Pages
      <li
        class="text-white"
      >
        <a
          class="p-2 text-white bg-gray-300"
          href="/?page=1"
        >
          1
        </a>
      </li>
      <li
        class="text-white"
      >
        <a
          class="p-2 "
          href="/?page=2"
        >
          2
        </a>
      </li>
      <li
        class="text-white"
      >
        <a
          class="p-2 "
          href="/?page=3"
        >
          3
        </a>
      </li>
      <li
        class="text-white"
      >
        <a
          class="p-2 "
          href="/?page=4"
        >
          4
        </a>
      </li>
      <li
        class="text-white"
      >
        <a
          class="p-2 "
          href="/?page=5"
        >
          5
        </a>
      </li>
      <li
        class="text-white"
      >
        <a
          class="p-2 "
          href="/?page=6"
        >
          6
        </a>
      </li>
      <li
        class="text-white"
      >
        <a
          class="p-2 "
          href="/?page=7"
        >
          7
        </a>
      </li>
      <li
        class="text-white"
      >
        <a
          class="p-2 "
          href="/?page=8"
        >
          8
        </a>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;

exports[`Should render the first page on load It should be render character list on page 2 1`] = `
<DocumentFragment>
  <div
    class="container mx-auto"
  >
    <ul
      class="flex flex-col gap-6 py-20"
    >
      <li
        class="flex flex-col shadow-md bg-gray-500 border rounded"
      >
        <a
          href="/character/1"
        >
          Luke Skywalker
        </a>
        <div>
          male
        </div>
        <div>
          https://swapi.dev/api/people/1/
        </div>
      </li>
      <li
        class="flex flex-col shadow-md bg-gray-500 border rounded"
      >
        <a
          href="/character/2"
        >
          C-3PO
        </a>
        <div>
          n/a
        </div>
        <div>
          https://swapi.dev/api/people/2/
        </div>
      </li>
      <li
        class="flex flex-col shadow-md bg-gray-500 border rounded"
      >
        <a
          href="/character/3"
        >
          R2-D2
        </a>
        <div>
          n/a
        </div>
        <div>
          https://swapi.dev/api/people/3/
        </div>
      </li>
      <li
        class="flex flex-col shadow-md bg-gray-500 border rounded"
      >
        <a
          href="/character/4"
        >
          Darth Vader
        </a>
        <div>
          male
        </div>
        <div>
          https://swapi.dev/api/people/4/
        </div>
      </li>
      <li
        class="flex flex-col shadow-md bg-gray-500 border rounded"
      >
        <a
          href="/character/5"
        >
          Leia Organa
        </a>
        <div>
          female
        </div>
        <div>
          https://swapi.dev/api/people/5/
        </div>
      </li>
      <li
        class="flex flex-col shadow-md bg-gray-500 border rounded"
      >
        <a
          href="/character/6"
        >
          Owen Lars
        </a>
        <div>
          male
        </div>
        <div>
          https://swapi.dev/api/people/6/
        </div>
      </li>
      <li
        class="flex flex-col shadow-md bg-gray-500 border rounded"
      >
        <a
          href="/character/7"
        >
          Beru Whitesun lars
        </a>
        <div>
          female
        </div>
        <div>
          https://swapi.dev/api/people/7/
        </div>
      </li>
      <li
        class="flex flex-col shadow-md bg-gray-500 border rounded"
      >
        <a
          href="/character/8"
        >
          R5-D4
        </a>
        <div>
          n/a
        </div>
        <div>
          https://swapi.dev/api/people/8/
        </div>
      </li>
      <li
        class="flex flex-col shadow-md bg-gray-500 border rounded"
      >
        <a
          href="/character/9"
        >
          Biggs Darklighter
        </a>
        <div>
          male
        </div>
        <div>
          https://swapi.dev/api/people/9/
        </div>
      </li>
      <li
        class="flex flex-col shadow-md bg-gray-500 border rounded"
      >
        <a
          href="/character/10"
        >
          Obi-Wan Kenobi
        </a>
        <div>
          male
        </div>
        <div>
          https://swapi.dev/api/people/10/
        </div>
      </li>
    </ul>
    <ul
      class="fixed bottom-0 flex gap-6 py-6 bg-gray-800 w-full"
    >
      Pages
      <li
        class="text-white"
      >
        <a
          class="p-2 "
          href="/?page=1"
        >
          1
        </a>
      </li>
      <li
        class="text-white"
      >
        <a
          class="p-2 text-white bg-gray-300"
          href="/?page=2"
        >
          2
        </a>
      </li>
      <li
        class="text-white"
      >
        <a
          class="p-2 "
          href="/?page=3"
        >
          3
        </a>
      </li>
      <li
        class="text-white"
      >
        <a
          class="p-2 "
          href="/?page=4"
        >
          4
        </a>
      </li>
      <li
        class="text-white"
      >
        <a
          class="p-2 "
          href="/?page=5"
        >
          5
        </a>
      </li>
      <li
        class="text-white"
      >
        <a
          class="p-2 "
          href="/?page=6"
        >
          6
        </a>
      </li>
      <li
        class="text-white"
      >
        <a
          class="p-2 "
          href="/?page=7"
        >
          7
        </a>
      </li>
      <li
        class="text-white"
      >
        <a
          class="p-2 "
          href="/?page=8"
        >
          8
        </a>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
